stdlib/domain.ml

```diff
@@ -1,11 +1,90 @@
-type t = int
-external spawn : (unit -> unit) -> t = "caml_domain_spawn"
+module Raw = struct
+  (* Low-level primitives provided by the runtime *)
+  type t = private int
+  external critical_adjust : int -> unit
+    = "caml_ml_domain_critical_section"
+  external interrupt : t -> unit
+    = "caml_ml_domain_interrupt"
+  external wait : unit -> unit = "caml_ml_domain_yield"
+  external spawn : (unit -> unit) -> t
+    = "caml_domain_spawn"
+  external self : unit -> t
+    = "caml_ml_domain_id"
+end
+
+module Sync = struct
+  let critical_section f =
+    Raw.critical_adjust (+1);
+    match f () with
+    | x -> Raw.critical_adjust (-1); x
+    | exception ex -> Raw.critical_adjust (-1); raise ex
+
+  let notify d = Raw.interrupt d
+  let wait () = Raw.wait ()
+end
+
+type id = Raw.t
+
+type 'a state =
+| Running
+| Joining of ('a, exn) result option ref * id
+| Finished of ('a, exn) result
+| Joined
+
+type 'a t =
+  { domain : Raw.t; state : 'a state Atomic.t }
+
+exception Retry
+let rec spin f =
+  try f () with Retry ->
+     (* fixme: spin more gently *)     
```

Comment:
Yes. The next feature I want to add is a couple of primitives for sleeping a defined amount of time or until an interrupt arrives.